pymc_bart/tree.py

```diff
@@ -267,33 +274,37 @@ def _traverse_tree(
         npt.NDArray[np.float_]
             Leaf node value or mean of leaf node values
         """
-        current_node = self.get_node(node_index)
-        if current_node.is_leaf_node():
-            if current_node.linear_params is None:
-                return np.array(current_node.value)
-
-            x = x[split_variable].item()
-            y_x = current_node.linear_params[0] + current_node.linear_params[1] * x
-            return np.array(y_x / m)
-
-        split_variable = current_node.idx_split_variable
-
-        if excluded is not None and current_node.idx_split_variable in excluded:
-            leaf_values: List[npt.NDArray[np.float_]] = []
-            leaf_n_values: List[int] = []
-            self._traverse_leaf_values(leaf_values, leaf_n_values, node_index)
-            return (
-                leaf_values[0].mean(axis=0) * leaf_n_values[0]
-                + leaf_values[1].mean(axis=0) * leaf_n_values[1]
-            )
-
-        if x[current_node.idx_split_variable] <= current_node.value:
-            next_node = get_idx_left_child(node_index)
-        else:
-            next_node = get_idx_right_child(node_index)
-        return self._traverse_tree(
-            x=x, m=m, node_index=next_node, split_variable=split_variable, excluded=excluded
-        )
+        stack = [(0, 1.0)]  # (node_index, prop) initial state
+        p_d = np.zeros(shape)
+        while stack:
+            node_index, weight = stack.pop()
+            node = self.get_node(node_index)
+            if node.is_leaf_node():
+                params = node.linear_params
+                if params is None:
+                    p_d += weight * node.value
+                else:
+                    # this produce nonsensical results
+                    p_d += weight * ((params[0] + params[1] * x[node.idx_split_variable]) / m)
```

Comment:
One quick (naive) question...when fitting the linear model, we use as covariates `X[idx_data_point, selected_predictor]` (see https://github.com/pymc-devs/pymc-bart/blob/main/pymc_bart/pgbart.py#L433) ... how does this map to `x[node.idx_split_variable]` aren't we missing the selected predictor specification?

Reply:
The reason I ask is because when using ` p_d += weight * node.value.mean()` I do get a nice pdp
![image](https://github.com/pymc-devs/pymc-bart/assets/22996444/843b831d-bc69-4821-a091-5ca2fbc16e84)

Reply:
what we call `selected_predictor` here `X[idx_data_point, selected_predictor]` is equivalent to what we call  `idx_split_variable]` here `x[node.idx_split_variable]`, in the first case `X` is a matrix and has many `idx_data_points` in the second case `x` is a vector or a single data_point.

By doing `node.value.mean()` we are using the mean of the stored/fitted values when doing `p_d += weight * ((params[0] + params[1] * x[node.idx_split_variable]) / m)` we are generating new predictions.  This seems to indicate the fitting is ok, but the predictions are not.